src/main/java/io/securecodebox/persistence/defectdojo/service/ProductTypeService.java

```diff
@@ -10,7 +10,7 @@
 import io.securecodebox.persistence.defectdojo.model.PaginatedResult;
 import io.securecodebox.persistence.defectdojo.model.ProductType;
 
-public class ProductTypeService extends GenericDefectDojoService<ProductType> {
+public final class ProductTypeService extends GenericDefectDojoService<ProductType> {
```

Comment:
This is a general advice to make API as private and immutable as possible to avoid accidental missuse. I make classes in general package private and final. Only, and only if, I later encounter that it is necessary to open such contraints, I do this explicitly. This practice helps me to avoid accidental complexity (e.g. inheritance hell).

Another example – not relevant for this particular example because it is a service and not a value object – is `equalTo()` and `hashCode()`. If your class is not final then you must handle [Liskov Substitution Principle](https://www.artima.com/articles/how-to-write-an-equality-method-in-java) to avoid possible memory leaks. That's quite complicated, so it makes sense to forbid deriving a model class to avoid this.

A very good book about this is _Effecitve Java_ from Joshua Bloch.